# Exercise 4.8: Grouping Data & Aggregating Variables agg()
#### by Claudia Lan Yee Chiu

### Importing libraries

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

### Importing Dataframes

In [2]:
# Assign path
path = r'C:\Users\clany\OneDrive\Documents\CareerFoundry\Chap_4_Pythons Fundamentals for Data Analysts\Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\clany\\OneDrive\\Documents\\CareerFoundry\\Chap_4_Pythons Fundamentals for Data Analysts\\Instacart Basket Analysis'

In [4]:
# Importing ords_prods_merged pkl from last exercise 4.7
df_ords_prods_merge = pd.read_pickle(os.path.join(path, 'Data', 'Prepared Data','orders_products_merged_task4_7.pkl'))

In [5]:
# Check imported dataframe
df_ords_prods_merge.head()

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0   2539329        1    prior                 1                   2   
1   2398795        1    prior                 2                   3   
2    473747        1    prior                 3                   3   
3   2254736        1    prior                 4                   4   
4    431534        1    prior                 5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   

   price_range_loc  price_range_loc     busiest_day  Busiest_days  \
0              NaN              NaN  Regularly busy   Normal days   
1              NaN              NaN  Regularly busy  Slowest days   
2              NaN              NaN  Regularly busy  Slowest days   
3              NaN              NaN      Least busy  Slowest days   
4              NaN              NaN      Least busy  Slowest days   

  busiest_period_of_day  
0        Average Orders  
1        Average Orders  
2           Most Orders  
3        Average Orders  
4           Most Orders

In [6]:
df_ords_prods_merge.shape

(32404859, 20)

In [7]:
# Creating a subset called df of the first one million entries as per exercise hands-on

df = df_ords_prods_merge[:1000000]

In [8]:
# Check the shape of newly created df
df.shape

(1000000, 20)

In [9]:
df.head(10)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0   2539329        1    prior                 1                   2   
1   2398795        1    prior                 2                   3   
2    473747        1    prior                 3                   3   
3   2254736        1    prior                 4                   4   
4    431534        1    prior                 5                   4   
5   3367565        1    prior                 6                   2   
6    550135        1    prior                 7                   1   
7   3108588        1    prior                 8                   1   
8   2295261        1    prior                 9                   1   
9   2550362        1    prior                10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered product_name  aisle_id  department_id  prices _merge  \
0          0         Soda        77              7     9.0   both   
1          1         Soda        77              7     9.0   both   
2          1         Soda        77              7     9.0   both   
3          1         Soda        77              7     9.0   both   
4          1         Soda        77              7     9.0   both   
5          1         Soda        77              7     9.0   both   
6          1         Soda        77              7     9.0   both   
7          1         Soda        77              7     9.0   both   
8          1         Soda        77              7     9.0   both   
9          1         Soda        77              7     9.0   both   

   price_range_loc  price_range_loc     busiest_day  Busiest_days  \
0              NaN              NaN  Regularly busy   Normal days   
1              NaN              NaN  Regularly busy  Slowest days   
2              NaN              NaN  Regularly busy  Slowest days   
3              NaN              NaN      Least busy  Slowest days   
4              NaN              NaN      Least busy  Slowest days   
5              NaN              NaN  Regularly busy   Normal days   
6              NaN              NaN  Regularly busy  Busiest days   
7              NaN              NaN  Regularly busy  Busiest days   
8              NaN              NaN  Regularly busy  Busiest days   
9              NaN              NaN      Least busy  Slowest days   

  busiest_period_of_day  
0        Average Orders  
1        Average Orders  
2           Most Orders  
3        Average Orders  
4           Most Orders  
5        Average Orders  
6           Most Orders  
7           Most Orders  
8           Most Orders  
9        Average Orders

### Grouping Data with pandas

#### First: Split the data into groups based on the given criteria
#### groupby() function: will group a given dataframe by a given column

In [10]:
# Use the groupby function first as example on the product_name

df.groupby('product_name')

#### Second: Aggregating  Data with agg()

In [12]:
# Calculate the mean of the ´order_number` column grouped by the department_id column to compare the average nummber of orders per user
# 1. Split the data into groups on department_id

df.groupby('department_id').agg({'number_of_orders': ['mean']})

number_of_orders
                          mean
department_id                 
4                    18.825780
7                    17.472355
13                   17.993423
14                   19.246334
16                   19.463012
17                   11.294069
19                   19.305237
20                   17.599636

In [13]:
# Aggregation conducted withouth the agg() function  - same result as above but different format

df.groupby('department_id')['number_of_orders'].mean()

department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: number_of_orders, dtype: float64

#### Performing Multiple Aggregations

In [14]:
# To produce multiple statistics at the same time, such as mean, minimum, maximum

df.groupby('department_id').agg({'number_of_orders': ['mean', 'min', 'max']})

number_of_orders        
                          mean min max
department_id                         
4                    18.825780   1  99
7                    17.472355   1  99
13                   17.993423   1  99
14                   19.246334   1  99
16                   19.463012   1  99
17                   11.294069   1  98
19                   19.305237   1  99
20                   17.599636   1  99

#### Third: Aggregrating Data with Transform() - creating flags for the data

In [16]:
# Create max_order column: three steps in one line of code
# 1. Split data into groups based on user_id, 2. Apply transform() on order_number to generate max orders of each user, 3. create max_order for the results 

df_ords_prods_merge['max_order'] = df_ords_prods_merge.groupby(['user_id'])['number_of_orders'].transform(np.max)

In [17]:
# Check the outcome via header

df_ords_prods_merge.head(15)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   2968173       15    prior                15                   1   
11   1870022       15    prior                17                   2   
12   1911383       15    prior                18                   2   
13   2715276       15    prior                21                   1   
14    487368       15    prior                22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered  ... aisle_id  department_id  prices  _merge  price_range_loc  \
0           0  ...       77              7     9.0    both              NaN   
1           1  ...       77              7     9.0    both              NaN   
2           1  ...       77              7     9.0    both              NaN   
3           1  ...       77              7     9.0    both              NaN   
4           1  ...       77              7     9.0    both              NaN   
5           1  ...       77              7     9.0    both              NaN   
6           1  ...       77              7     9.0    both              NaN   
7           1  ...       77              7     9.0    both              NaN   
8           1  ...       77              7     9.0    both              NaN   
9           1  ...       77              7     9.0    both              NaN   
10          0  ...       77              7     9.0    both              NaN   
11          1  ...       77              7     9.0    both              NaN   
12          1  ...       77              7     9.0    both              NaN   
13          1  ...       77              7     9.0    both              NaN   
14          1  ...       77              7     9.0    both              NaN   

   price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0              NaN  Regularly busy   Normal days        Average Orders   
1              NaN  Regularly busy  Slowest days        Average Orders   
2              NaN  Regularly busy  Slowest days           Most Orders   
3        

In [18]:
# View 100 rows via head() function
df_ords_prods_merge.head(100)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
..       ...      ...      ...               ...                 ...   
95   3226575      360    prior                 1                   5   
96   1469869      377    prior                 3                   5   
97   1927023      387    prior                 2                   4   
98    858092      420    prior                 4                   1   
99    119314      420    prior                 5                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
..                ...                     ...         ...                ...   
95                 12                     NaN         196                  1   
96                 17                     3.0         196                  9   
97                 10                    22.0         196                  3   
98                 19                    30.0         196                  2   
99                 19                    16.0         196                  2   

    reordered  ... aisle_id  department_id  prices  _merge  price_range_loc  \
0           0  ...       77              7     9.0    both              NaN   
1           1  ...       77              7     9.0    both              NaN   
2           1  ...       77              7     9.0    both              NaN   
3           1  ...       77              7     9.0    both              NaN   
4           1  ...       77              7     9.0    both              NaN   
..        ...  ...      ...            ...     ...     ...              ...   
95          0  ...       77              7     9.0    both              NaN   
96          0  ...       77              7     9.0    both              NaN   
97          0  ...       77              7     9.0    both              NaN   
98          0  ...       77              7     9.0    both              NaN   
99          1  ...       77              7     9.0    both              NaN   

   price_range_loc     busiest_day  Busiest_days busiest_period_of_day  \
0              NaN  Regularly busy   Normal days        Average Orders   
1              NaN  Regularly busy  Slowest days        Average Orders   
2              NaN  Regularly busy  Slowest days           Most Orders   
3              NaN      Least busy  Slowest days        Average Orders   
4              NaN      Least busy  Slowest days           Most Orders   
..             ...             ...           ...                   ...   
95             NaN  Regularly busy   Normal days           Most Orders   
96             NaN  Regularly busy   Normal days        Average Orders   
97             NaN      Least busy  Slowest days           Most Orders   
98             NaN  Regularly busy  Busiest days        Average Orders   
99             NaN  Regularly busy  Slowest days        Average Orders   

   max_order  
0         10  
1         10  
2         10  
3         10  
4         10  
..       ...  
95         3  
96         3  
97         8  
98        22  
99        22  

[100 rows x 21 columns]

In [19]:
# Request pandas NOT to assign any option regarding the maximum number of rows to display

pd.options.display.max_rows = None

In [ ]:
# View all 100 rows via head() function
df_ords_prods_merge.head(100)

###### Cleared output of above to reduce the size of notebook!

#### Deriving Columns with loc()

##### Create the flag with following criteria:

##### 1. If the maximum orders the user has made is over 40, then the customer will be labeled a “Loyal customer.”
##### 2. If the maximum orders the user has made is over 10 but less than or equal to 40, then the customer will be labeled a “Regular customer.”
##### 3. If the maximum orders the user has made is less than or equal to 10, then the customer will be labeled a “New customer.”

In [22]:
# Condition #1 above
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal Customer'

In [23]:
# Condition #2 above
df_ords_prods_merge.loc[(df_ords_prods_merge['max_order'] <= 40) & (df_ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular Customer'

In [24]:
# Condition #3 above
df_ords_prods_merge.loc[df_ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New Customer'

In [26]:
# Print frequency of new ´loyalty_flag` using value_counts
df_ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: loyalty_flag, dtype: int64

In [32]:
# To view only specific columns of a dataframe

# df_ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_orders']].head(5)

# Received 
# ---------------------------------------------------------------------------
# MemoryError                               Traceback (most recent call last)
# Cell In[29], line 3
#       1 # To view only specific columns of a dataframe
# ----> 3 df_ords_prods_merge[['user_id', 'loyalty_flag', 'number_of_orders']].head(5)
# File <__array_function__ internals>:200, in concatenate(*args, **kwargs)

# MemoryError: Unable to allocate 2.66 GiB for an array with shape (11, 32404859) and data type int64

# Task 4.8 

#### Step 1: Create a new notebook for this task. Be sure to import the relevant libraries, along with your ords_prods_merge dataframe, which should include your newly derived columns from the previous Exercise
##### Done

#### Step 2: In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [33]:
# First step: use ´groupby` function on the ´product_name´

df_ords_prods_merge.groupby('product_name')

In [34]:
# Second step: split the data into groups by "department_id" and then calculate the mean for each "number_of_orders" column
df_ords_prods_merge.groupby('department_id').agg({'number_of_orders': ['mean']})

number_of_orders
                          mean
department_id                 
1                    15.457838
2                    17.277920
3                    17.170395
4                    17.811403
5                    15.215751
6                    16.439806
7                    17.225802
8                    15.340650
9                    15.895474
10                   20.197148
11                   16.170638
12                   15.887671
13                   16.583536
14                   16.773669
15                   16.165037
16                   17.665606
17                   15.694469
18                   19.310397
19                   17.177343
20                   16.473447
21                   22.902379

#### Step 3: Analyze the result. How do the results for the entire dataframe differ from those of the subset? 

##### The mean of the subset is greater than the one of the entire dataset (except for department_id: 17). 
##### Also, the minimum value of the subset is 11.3 while it is much greater in the entire df (15.2). The maximum value of the entire df is also much higher (23).

#### Step 4: Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions

In [35]:
# Done: see output 26 above

#### Step 5:  Determine whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers

In [36]:
# Check the data for accuracy
df_ords_prods_merge.head(100)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   2968173       15    prior                15                   1   
11   1870022       15    prior                17                   2   
12   1911383       15    prior                18                   2   
13   2715276       15    prior                21                   1   
14    487368       15    prior                22                   1   
15   2293453       19    prior                 2                   5   
16   1973799       19    prior                 5                   6   
17    532817       19    prior                 7                   4   
18   1573906       21    prior                10                   3   
19   1593000       31    prior                10                   3   
20   2231262       31    prior                17                   3   
21   2580647       43    prior                 6                   4   
22   2187180       43    prior                 9                   4   
23   2497897       52    prior                 1                   1   
24   1318871       52    prior                 2                   1   
25   1261384       52    prior                 3                   1   
26    580568       52    prior                 5                   2   
27   2428073       52    prior                 7                   2   
28    342306       52    prior                 9                   1   
29   1520353       52    prior                10                   2   
30   2542086       52    prior                11                   2   
31    944694       52    prior                12                   1   
32   1268191       52    prior                15                   2   
33   1122089       52    prior                17                   2   
34   1498922       52    prior                18                   2   
35    180919       52    prior                19                   3   
36   2695875       52    prior                21                   5   
37   3205960       67    prior                 1                   3   
38    970922       67    prior                 2                   1   
39    596585       67    prior                 3                   1   
40   2037087       67    prior                 4                   2   
41   2474111       67    prior                 5                   3   
42   1568481       67    prior                 6                   3   
43   2048527       67    prior                 8                   2   
44   2389031       67    prior                 9                   2   
45   2378562       67    prior                10                   1   
46   1515698       67    prior                11                   1   
47   1928050       67    prior                12                   2   
48   1949274       67    prior                13                   1   
49    551869       67    prior                14                   1   
50    768572       67    prior                16                   4   
51    150528       67    prior                18                   5   
52     29758       67    prior                19                   0   
53   1558976       67    prior                21                   1   
54    863974       67    prior              

In [37]:
# Basic statistics of df_orders_products_merged by 'loyalty_flag' for each loyalty category

df_ords_prods_merge.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal Customer    10.386336  1.0  99999.0
New Customer      13.294670  1.0  99999.0
Regular Customer  12.495717  1.0  99999.0

##### Prices of products purchased by loyal customers are much lower than new and regular customers. Interesting to note that new customers purchase in average the highest priced products.

#### Step 6: Create a spending flag for each user based on the average price across all their orders using the following criteria:
#### If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
#### If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [38]:
# Create a new column ´average_price´ and aggregate with transform() using numpy mean

df_ords_prods_merge['average_price']=df_ords_prods_merge.groupby(['user_id'])['prices'].transform(np.mean)

In [40]:
# Check output
df_ords_prods_merge.head(25)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   2968173       15    prior                15                   1   
11   1870022       15    prior                17                   2   
12   1911383       15    prior                18                   2   
13   2715276       15    prior                21                   1   
14    487368       15    prior                22                   1   
15   2293453       19    prior                 2                   5   
16   1973799       19    prior                 5                   6   
17    532817       19    prior                 7                   4   
18   1573906       21    prior                10                   3   
19   1593000       31    prior                10                   3   
20   2231262       31    prior                17                   3   
21   2580647       43    prior                 6                   4   
22   2187180       43    prior                 9                   4   
23   2497897       52    prior                 1                   1   
24   1318871       52    prior                 2                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   
20                 11                     8.0         196                 14   
21                 16                     4.0         196                  9   
22                 12                     3.0         196                  1   
23                  9                     NaN         196                  4   
24                 10                     7.0         196                  5   

    reordered  ... prices  _merge   price_range_lo

In [41]:
# Create ´Spending Flag´ column and assign labels ´Low spender´ for prices under 10

df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] < 10, 'spending_flag'] = 'Low spender'

In [42]:
# Create ´Spending Flag´ column and assign labels ´High spender´ for prices equal or above 10

df_ords_prods_merge.loc[df_ords_prods_merge['average_price'] >= 10, 'spending_flag'] = 'High spender'

In [43]:
# Check output

df_ords_prods_merge.head()

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0   2539329        1    prior                 1                   2   
1   2398795        1    prior                 2                   3   
2    473747        1    prior                 3                   3   
3   2254736        1    prior                 4                   4   
4    431534        1    prior                 5                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   

   reordered  ... _merge   price_range_loc  price_range_loc     busiest_day  \
0          0  ...   both               NaN              NaN  Regularly busy   
1          1  ...   both               NaN              NaN  Regularly busy   
2          1  ...   both               NaN              NaN  Regularly busy   
3          1  ...   both               NaN              NaN      Least busy   
4          1  ...   both               NaN              NaN      Least busy   

   Busiest_days busiest_period_of_day  max_order  loyalty_flag average_price  \
0   Normal days        Average Orders         10  New Customer      6.367797   
1  Slowest days        Average Orders         10  New Customer      6.367797   
2  Slowest days           Most Orders         10  New Customer      6.367797   
3  Slowest days        Average Orders         10  New Customer      6.367797   
4  Slowest days           Most Orders         10  New Customer      6.367797   

  spending_flag  
0   Low spender  
1   Low spender  
2   Low spender  
3   Low spender  
4   Low spender  

[5 rows x 24 columns]

In [44]:
# Check the frequency of new column ´spending flag´

df_ords_prods_merge['spending_flag'].value_counts(dropna = False)

Low spender     31770614
High spender      634245
Name: spending_flag, dtype: int64

#### Step 7: Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. The criteria for the flag should be as follows:
If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [45]:
# Create a new column ´median_prior_order´ and aggregate with transform() using numpy median

df_ords_prods_merge['median_prior_order']=df_ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform(np.mean)

In [46]:
# Check output

df_ords_prods_merge.head(25)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   2968173       15    prior                15                   1   
11   1870022       15    prior                17                   2   
12   1911383       15    prior                18                   2   
13   2715276       15    prior                21                   1   
14    487368       15    prior                22                   1   
15   2293453       19    prior                 2                   5   
16   1973799       19    prior                 5                   6   
17    532817       19    prior                 7                   4   
18   1573906       21    prior                10                   3   
19   1593000       31    prior                10                   3   
20   2231262       31    prior                17                   3   
21   2580647       43    prior                 6                   4   
22   2187180       43    prior                 9                   4   
23   2497897       52    prior                 1                   1   
24   1318871       52    prior                 2                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   
20                 11                     8.0         196                 14   
21                 16                     4.0         196                  9   
22                 12                     3.0         196                  1   
23                  9                     NaN         196                  4   
24                 10                     7.0         196                  5   

    reordered  ...  price_range_loc  price_range_l

In [47]:
# Create new column ´freqency flag´ based on median days since prior order and label ´Non-frequent Customer´ when higher than 20
df_ords_prods_merge.loc[df_ords_prods_merge['median_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [49]:
# Create new column ´freqency flag´ based on median days since prior order and label ´Regular Customer´ when higher than 10 and lower or equal 20
df_ords_prods_merge.loc[(df_ords_prods_merge['median_prior_order'] > 10) & (df_ords_prods_merge['median_prior_order'] <= 20), 'frequency_flag'] = 'Regular customer'

In [50]:
# Create new column ´freqency flag´ based on median days since prior order and label ´Frequent Customer´ is lower than or equal to 10
df_ords_prods_merge.loc[df_ords_prods_merge['median_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [51]:
# Check output

df_ords_prods_merge.head(21)

order_id  user_id eval_set  number_of_orders  orders_day_of_week  \
0    2539329        1    prior                 1                   2   
1    2398795        1    prior                 2                   3   
2     473747        1    prior                 3                   3   
3    2254736        1    prior                 4                   4   
4     431534        1    prior                 5                   4   
5    3367565        1    prior                 6                   2   
6     550135        1    prior                 7                   1   
7    3108588        1    prior                 8                   1   
8    2295261        1    prior                 9                   1   
9    2550362        1    prior                10                   4   
10   2968173       15    prior                15                   1   
11   1870022       15    prior                17                   2   
12   1911383       15    prior                18                   2   
13   2715276       15    prior                21                   1   
14    487368       15    prior                22                   1   
15   2293453       19    prior                 2                   5   
16   1973799       19    prior                 5                   6   
17    532817       19    prior                 7                   4   
18   1573906       21    prior                10                   3   
19   1593000       31    prior                10                   3   
20   2231262       31    prior                17                   3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   
15                 14                     6.0         196                  3   
16                 12                     8.0         196                 15   
17                 17                     6.0         196                  1   
18                 10                     6.0         196                  2   
19                  8                     7.0         196                 17   
20                 11                     8.0         196                 14   

    reordered  ... price_range_loc     busiest_day  Busiest_days  \
0           0  ...             NaN  Regularly busy   Normal days   
1           1  ...             NaN  Regularly busy  Slowest days   
2           1  ...             NaN  Regularly busy  Slowest days   
3           1  ...             NaN      Least busy  Slowest days   
4           1  ...             NaN      Least busy  Slowest days   
5           1  ...             NaN  Regularly busy   Normal days   
6           1  ...             NaN  Regularly busy  Busiest days   
7           1  ...             NaN  Regularly busy  Busiest days   
8           1  ...             NaN  Regularly 

In [52]:
# Check the frequency of new column ´frequency flag´

df_ords_prods_merge['frequency_flag'].value_counts(dropna = False)

Frequent customer        16468582
Regular customer         12521159
Non-frequent customer     3415113
NaN                             5
Name: frequency_flag, dtype: int64

In [63]:
# Check the null customer above

df_ords_prods_merge[df_ords_prods_merge['frequency_flag'].isnull()][['user_id', 'days_since_prior_order', 'frequency_flag']]

user_id  days_since_prior_order frequency_flag
13645692   159838                     NaN            NaN
17251990   159838                     NaN            NaN
17622767   159838                     NaN            NaN
24138593   159838                     NaN            NaN
25880002   159838                     NaN            NaN

##### The NaN values seem to be from a new customer with user_id 159838 with no prior order, meaning this is his first order, hence the NaN is justified

#### Last Step of this task: Export your dataframe as a pickle file and store it correctly in your “Prepared Data” folder.

In [65]:
# Export as pickle
df_ords_prods_merge.to_pickle(os.path.join(path, 'Data', 'Prepared Data', 'orders_products_aggregated.pkl'))